src/shims/unix/fs.rs

```diff
@@ -1721,6 +1722,129 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
         this.set_last_error(enotty)?;
         Ok(0)
     }
+
+    fn mkstemp(&mut self, template_op: &OpTy<'tcx, Tag>) -> InterpResult<'tcx, i32> {
+        use rand::seq::SliceRandom;
+
+        // POSIX defines the template string.
+        const TEMPFILE_TEMPLATE_STR: &str = "XXXXXX";
+        // POSIX defines the minimum number of attempts before failure.
+        // <https://github.com/lattera/glibc/blob/895ef79e04a953cac1493863bcae29ad85657ee1/sysdeps/posix/tempname.c#L203>
+        const TEMPFILE_ATTEMPTS_MIN: u32 = 62 * 62 * 62;
+
+        let this = self.eval_context_mut();
+        this.assert_target_os_is_unix("mkstemp");
+
+        // Get the raw bytes from the template. This is passed to us from `std` as an `OsString`.
+        // These bytes represent a string in _target_ OS encoding, which may or may not match
+        // the _host_ OS encoding.
+        let template_ptr = this.read_pointer(template_op)?;
+        let mut template = this.eval_context_ref().read_c_str(template_ptr)?.to_owned();
+        let template_bytes = template.as_mut_slice();
+
+        // Reject if isolation is enabled.
+        if let IsolatedOp::Reject(reject_with) = this.machine.isolated_op {
+            this.reject_in_isolation("`mkstemp`", reject_with)?;
+            let eacc = this.eval_libc("EACCES")?;
+            this.set_last_error(eacc)?;
+            return Ok(-1);
+        }
+
+        // Get the bytes of the suffix we expect in _target_ encoding.
+        let suffix_bytes = TEMPFILE_TEMPLATE_STR.as_bytes();
+
+        // At this point we have one `&[u8]` that represents the template and one `&[u8]`
+        // that represents the suffix, both with the same (_target_) encoding.
+        // Critically, the encoding may differ from the host / where _this_ code is running!
```

Comment:
I am not entirely sure what you mean here. Which "encoding" are you referring to?

Reply:
The character encoding of the `OSString`, which is different on Windows and Unix AFAICT.